File: src/middleware/rate-limit.ts
import type { MiddlewareHandler } from 'hono';
import { HTTP_STATUS } from '@/config/constants';
import type { Account } from '@/db/schema';
import { logger } from '@/utils/logger';

interface RateLimitEntry {
  timestamps: number[];
  lastAccess: number;
}

const rateLimitStore = new Map<string, RateLimitEntry>();
const MAX_ENTRIES = 10000;
const CLEANUP_INTERVAL_MS = 60 * 1000;
const ENTRY_TTL_MS = 5 * 60 * 1000;

let lastCleanup = Date.now();

function cleanupStaleEntries(): void {
  const now = Date.now();
  if (now - lastCleanup < CLEANUP_INTERVAL_MS) {
    return;
  }
  lastCleanup = now;

  for (const [key, entry] of rateLimitStore.entries()) {
    if (now - entry.lastAccess > ENTRY_TTL_MS) {
      rateLimitStore.delete(key);
    }
  }

  if (rateLimitStore.size > MAX_ENTRIES) {
    const entries = [...rateLimitStore.entries()].sort((a, b) => a[1].lastAccess - b[1].lastAccess);
    const deleteCount = Math.floor(entries.length * 0.2);
    for (let i = 0; i < deleteCount; i++) {
      const entry = entries[i];
      if (entry) {
        rateLimitStore.delete(entry[0]);
      }
    }
  }
}

function cleanupOldTimestamps(entry: RateLimitEntry, windowMs: number): void {
  const now = Date.now();
  entry.timestamps = entry.timestamps.filter((ts) => now - ts < windowMs);
}

function checkRateLimit(
  accountId: string,
  limit: number
): { allowed: boolean; remaining: number; resetAt: number } {
  const windowMs = 60 * 1000;
  const now = Date.now();

  cleanupStaleEntries();

  let entry = rateLimitStore.get(accountId);
  if (!entry) {
    entry = { timestamps: [], lastAccess: now };
    rateLimitStore.set(accountId, entry);
  }

  entry.lastAccess = now;
  cleanupOldTimestamps(entry, windowMs);

  const remaining = Math.max(0, limit - entry.timestamps.length);
  const firstTimestamp = entry.timestamps[0];
  const resetAt =
    firstTimestamp !== undefined
      ? Math.ceil((firstTimestamp + windowMs) / 1000)
      : Math.ceil((now + windowMs) / 1000);

  if (entry.timestamps.length >= limit) {
    return { allowed: false, remaining: 0, resetAt };
  }

  entry.timestamps.push(now);
  return { allowed: true, remaining: remaining - 1, resetAt };
}

export function rateLimitMiddleware(): MiddlewareHandler {
  return async (c, next) => {
    // Get account from context (set by auth middleware)
    const account = c.get('account') as Account | undefined;

    if (!account) {
      // No account = no rate limiting (will fail auth anyway)
      await next();
      return;
    }

    const limit = account.rateLimitPerMinute ?? 60;
    const { allowed, remaining, resetAt } = checkRateLimit(account.id, limit);

    // Set rate limit headers
    c.header('X-RateLimit-Limit', limit.toString());
    c.header('X-RateLimit-Remaining', remaining.toString());
    c.header('X-RateLimit-Reset', resetAt.toString());

    if (!allowed) {
      logger.warn('Rate limit exceeded', { accountId: account.id });
      c.header('Retry-After', '60');
      return c.json({ error: 'Rate limit exceeded' }, HTTP_STATUS.TOO_MANY_REQUESTS);
    }

    await next();
  };
}

File: internal/middleware/ratelimit.go
package middleware

import (
	"net/http"
	"strconv"
	"sync"
	"time"

	"github.com/rs/zerolog/log"

	"github.com/openclaw/relay-server-go/internal/config"
)

const (
	maxEntries      = 10000
	cleanupInterval = time.Minute
	entryTTL        = 5 * time.Minute
	windowDuration  = time.Minute
)

type rateLimitEntry struct {
	timestamps []time.Time
	lastAccess time.Time
}

type RateLimiter struct {
	mu          sync.RWMutex
	store       map[string]*rateLimitEntry
	lastCleanup time.Time
}

func NewRateLimiter() *RateLimiter {
	return &RateLimiter{
		store:       make(map[string]*rateLimitEntry),
		lastCleanup: time.Now(),
	}
}

func (rl *RateLimiter) cleanup() {
	now := time.Now()
	if now.Sub(rl.lastCleanup) < cleanupInterval {
		return
	}
	rl.lastCleanup = now

	for key, entry := range rl.store {
		if now.Sub(entry.lastAccess) > entryTTL {
			delete(rl.store, key)
		}
	}

	if len(rl.store) > maxEntries {
		oldest := make([]string, 0, len(rl.store)/5)
		for key := range rl.store {
			oldest = append(oldest, key)
			if len(oldest) >= len(rl.store)/5 {
				break
			}
		}
		for _, key := range oldest {
			delete(rl.store, key)
		}
	}
}

func (rl *RateLimiter) Check(accountID string, limit int) (allowed bool, remaining int, resetAt int64) {
	rl.mu.Lock()
	defer rl.mu.Unlock()

	rl.cleanup()

	now := time.Now()
	windowStart := now.Add(-windowDuration)

	entry, exists := rl.store[accountID]
	if !exists {
		entry = &rateLimitEntry{
			timestamps: make([]time.Time, 0),
			lastAccess: now,
		}
		rl.store[accountID] = entry
	}

	entry.lastAccess = now

	filtered := entry.timestamps[:0]
	for _, ts := range entry.timestamps {
		if ts.After(windowStart) {
			filtered = append(filtered, ts)
		}
	}
	entry.timestamps = filtered

	remaining = limit - len(entry.timestamps)
	if remaining < 0 {
		remaining = 0
	}

	if len(entry.timestamps) > 0 {
		resetAt = entry.timestamps[0].Add(windowDuration).Unix()
	} else {
		resetAt = now.Add(windowDuration).Unix()
	}

	if len(entry.timestamps) >= limit {
		return false, 0, resetAt
	}

	entry.timestamps = append(entry.timestamps, now)
	return true, remaining - 1, resetAt
}

type RateLimitMiddleware struct {
	limiter *RateLimiter
}

func NewRateLimitMiddleware() *RateLimitMiddleware {
	return &RateLimitMiddleware{
		limiter: NewRateLimiter(),
	}
}

func (m *RateLimitMiddleware) Handler(next http.Handler) http.Handler {
	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		account := GetAccount(r.Context())
		if account == nil {
			next.ServeHTTP(w, r)
			return
		}

		limit := account.RateLimitPerMin
		if limit <= 0 {
			limit = config.DefaultRateLimitPerMin
		}

		allowed, remaining, resetAt := m.limiter.Check(account.ID, limit)

		w.Header().Set("X-RateLimit-Limit", strconv.Itoa(limit))
		w.Header().Set("X-RateLimit-Remaining", strconv.Itoa(remaining))
		w.Header().Set("X-RateLimit-Reset", strconv.FormatInt(resetAt, 10))

		if !allowed {
			log.Warn().Str("accountId", account.ID).Msg("rate limit exceeded")
			w.Header().Set("Retry-After", "60")
			writeJSON(w, http.StatusTooManyRequests, map[string]string{
				"error": "Rate limit exceeded",
			})
			return
		}

		next.ServeHTTP(w, r)
	})
}
